loadbalancer/loadbalancer_test.go

```diff
@@ -0,0 +1,633 @@
+			//ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			//defer cancel()
+			tn := setupTestNet(ctx, t, peers)
+			testClock := clock.NewMock()
+			testClock.Set(startTime)
+			lb := newLoadBalancer(ctx, tn.loadBalancer, nil, testClock)
+			lb.Start()
+			// register routes
+			routingResponse := tn.sendRoutingRequest(func(w io.Writer) error {
+				return messages.WriteRoutingRequest(w, messages.RoutingKindNew, testProtocols)
+			})
+			require.Equal(t, &messages.RoutingResponse{
+				Code:   messages.ResponseOk,
+				Expiry: &expiry,
+			}, routingResponse)
+
+			if testCase.registerHandler {
+				handler := func(s network.Stream) {
+					request, err := io.ReadAll(s)
+					require.NoError(t, err)
+					require.Equal(t, "request", string(request))
+					_, err = s.Write([]byte("response"))
+					require.NoError(t, err)
+					s.Close()
+				}
+				tn.publicNode.SetStreamHandler(testProtocols[1], handler)
+				tn.publicNode.SetStreamHandler(otherProtocol, handler)
+			}
+			response, s := tn.openOutboundForwardingRequest(testCase.write)
+			require.Equal(t, testCase.expectedResponse, response)
+			if response.Code == messages.ResponseOk {
+				_, err := s.Write([]byte("request"))
+				require.NoError(t, err)
+				s.CloseWrite()
+				streamResponse, err := io.ReadAll(s)
+				require.NoError(t, err)
+				require.Equal(t, "response", string(streamResponse))
+				s.Close()
+			}
+			err := lb.Close()
+			require.NoError(t, err)
+		})
+	}
+}
+
+func TestInboundForwarding(t *testing.T) {
+	testProtocols := []protocol.ID{"applesauce/mcgee", "test/me"}
+	otherProtocol := protocol.ID("noprotocol/here")
+	startTime := time.Date(2001, 12, 1, 0, 0, 0, 0, time.UTC)
+	expiry := time.Unix(0, startTime.Add(DefaultDuration).UnixNano())
+	ctx := context.Background()
+	peers := makePeers(t)
+	testCases := []struct {
+		name             string
+		protocols        []protocol.ID
+		willErrorOpening bool
+		willErrorReading bool
+		registerHandler  bool
+		rejectResponse   bool
+	}{
+		{
+			name:             "simple success",
+			protocols:        testProtocols,
+			willErrorOpening: false,
+			willErrorReading: false,
+			registerHandler:  true,
+		},
+		{
+			name:             "error - protocol not registered",
+			protocols:        []protocol.ID{otherProtocol},
+			willErrorOpening: true,
+			willErrorReading: false,
+			registerHandler:  true,
+		},
+		{
+			name:             "error - connecting on forwarding protocol error",
+			protocols:        testProtocols,
+			willErrorOpening: false,
+			willErrorReading: true,
+			registerHandler:  false,
+		},
+		{
+			name:             "error - response rejected",
+			protocols:        testProtocols,
+			willErrorOpening: false,
+			willErrorReading: true,
+			registerHandler:  true,
+			rejectResponse:   true,
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			//ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			//defer cancel()
```

Comment:
were they taking too long or did you forget to uncomment?

Reply:
it would appear so